### PR TITLE
DAOS-11695 test: Add test users to CI env

### PR DESCRIPTION
**Pull request by daltonbohning**
Skip-test: true
Skip-build: true
Doc-only: true
Test-tag: test_harness_pre_launch_users

Required-githooks: true

Signed-off-by: Dalton Bohning <dalton.bohning@intel.com>

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it
      should be force landed and forced-landing tag is set.
* [x] Commit messages follows the guidelines outlined
      [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [x] You are the appropriate gatekeeper to be landing the patch.
* [x] The PR has 2 reviews by people familiar with the code.
* [x] Any appropriate watchers have had a chance to review the PR.
* [x] Review comments are sufficiently resolved, particularly by prior reviewers that requested
      changes.
* [x] Githooks were used. If not, there is a sufficient reason to move forward.  Check copyrights
      if githooks were not used.
* [x] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future
      PRs.
* [x] All builds have passed.  Check non-required builds to ensure they don't indicate any problem
      such as a compiler warning on extraneous platforms.
* [x] No new NLT or valgrind warnings.  Check the classic view. This step should only matter if
      the build requires force landing.
* [x] Ensure sufficent testing is done.   Check feature pragmas and test tags. Check that tests
      skipped for the ticket are run and now pass with the changes.
* [x] Quick-build or Quick-functional is not used.
* [x] Pay attention to PRs that may affect compatibility between versions and ensure it has been
      addressed.
* [x] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If
      it is a release branch, does it have merge approval in the JIRA ticket.
* [x] Check the commit message when landing. Check the standard
      [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit
      it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by daltonbohning** [comment]
@brianjmurrell @phender Before running this on CI nodes, I'd like to sanity check the commands. These are the commands ultimately ran on each test node:
```
# Create two groups
groupadd daos_test_groupx -r
groupadd daos_test_groupy -r

# Delete users if pre-existing
if $(id -u daos_test_userx1 &> /dev/null); then userdel -f -r daos_test_userx1; fi
if $(id -u daos_test_userx2 &> /dev/null); then userdel -f -r daos_test_userx2; fi
if $(id -u daos_test_usery1 &> /dev/null); then userdel -f -r daos_test_usery1; fi

# Create users
useradd -m -g daos_test_groupx daos_test_userx1
useradd -m -g daos_test_groupx daos_test_userx2
useradd -m -g daos_test_groupy daos_test_usery1
```

File: src/tests/ftest/user_setup.py
Comment (phender) [inline src/tests/ftest/user_setup.py:125]: This should either default to empty list (so you don't have a problem on line 67) or the progarm should quit if no users are specified:
```suggestion
    users = args.users.split(',') if args.users else []
```

Alternatively, the string `split()` could be avoided by defining the `--user` argument with `nargs="*"` and then `required=True` should ensure at least one user was provided.

File: src/tests/ftest/user_setup.py
Comment (phender) [inline src/tests/ftest/user_setup.py:110]: Should this have `required=True` set?

File: src/tests/ftest/user_setup.py
Comment (phender) [inline src/tests/ftest/user_setup.py:113]: Should this be required too or is there a case where it makes sense to create a group without creating the user?

File: src/tests/ftest/user_setup.py
Comment (phender) [inline src/tests/ftest/user_setup.py:101]: Optionally the `default=get_local_host()` could be used here with the benefit of appearing in the help

File: src/tests/ftest/scripts/main.sh
Comment (phender) [inline src/tests/ftest/scripts/main.sh:197]: This should not be required
```suggestion
```

File: src/tests/ftest/scripts/main.sh
Comment (phender) [inline src/tests/ftest/scripts/main.sh:204]: This should not be required
```suggestion
```

File: src/tests/ftest/scripts/main.sh
Comment (phender) [inline src/tests/ftest/scripts/main.sh:194]: Optional readability improvement:
```suggestion
if ! ./user_setup.py -n "${TEST_NODES}" -g daos_test_group_x -u daos_test_user_x1,daos_test_user_x2 -s; then
```

File: src/tests/ftest/scripts/main.sh
Comment (phender) [inline src/tests/ftest/scripts/main.sh:201]: Optional readability improvement:
```suggestion
if ! ./user_setup.py -n "${TEST_NODES}" -g daos_test_group_y -u daos_test_user_y1 -s; then
```

**Review by phender — CHANGES_REQUESTED** [review]

File: src/tests/ftest/user_setup.py
Comment (daltonbohning) [reply src/tests/ftest/user_setup.py:110]: Maybe. At this point, I'm mostly trying to make sure I have the commands in order, then I plan to make the script itself robust

File: src/tests/ftest/user_setup.py
Comment (daltonbohning) [reply src/tests/ftest/user_setup.py:113]: `-d` is for when we don't want the user's home directory to be the default. I found this necessary on boro, since I don't have permission to create users in `/home` (it's shared across the system). So I use `-d /tmp`.
In CI, I don't what we should do

File: src/tests/ftest/user_setup.py
Comment (daltonbohning) [reply src/tests/ftest/user_setup.py:101]: Thanks. I'll clean up the interface before moving out of draft

File: src/tests/ftest/scripts/main.sh
Comment (daltonbohning) [reply src/tests/ftest/scripts/main.sh:194]: Agree. Thanks

File: src/tests/ftest/user_setup.py
Comment (ashleypittman) [reply src/tests/ftest/user_setup.py:125]: nargs="*" means a list of zero or more, nargs="+" means a list of one or more which I think is what you're looking for here.

File: src/tests/ftest/user_setup.py
Comment (ashleypittman) [inline src/tests/ftest/user_setup.py:15]: I'm somewhat familiar with this error but don't know the solution.

All other python files in src/test/ftest/*.py which import from util hit the same problem, and I've had to suppress it.  The check passes for me locally but fails in github actions. I think part of the reason is there is a python module called "util" but from memory it's present both in my setup and in GHA so that doesn't entirely explain it.

Anyway, you should just suppress this error for now as not a regression and I'll take another look next time I'm doing a parse of the pylint code.


https://github.com/daos-stack/daos/blob/master/src/tests/ftest/config_file_gen.py#L8-L9
https://github.com/daos-stack/daos/blob/master/src/tests/ftest/launch.py#L30
https://github.com/daos-stack/daos/blob/master/src/tests/ftest/process_core_files.py#L14
https://github.com/daos-stack/daos/blob/master/src/tests/ftest/slurm_setup.py#L8

File: src/tests/ftest/user_setup.py
Comment (ashleypittman) [inline src/tests/ftest/user_setup.py:50]: This file is python3 right, so should be using f-strings?

File: src/tests/ftest/user_setup.py
Comment (ashleypittman) [inline src/tests/ftest/user_setup.py:91]: nitpick, this code could in theory return a positive int in a way that wouldn't cause a failure.  The exit code of a program is a signed char with negative values reserved for signals so only 1-127 are valid "failure" return values for this function.  If this function returned 256 for example then when cast to a signed int it would be 0 and the check would pass.  Likewise if it was 501 then that would translate to -11 or SEGV.

This might seem academic and I'm not requesting changes here but we've seen exactly this issue in IOR where we had a data corruption issue where every byte was wrong and as the buffer was 2^N in size the error count and therefore exit code was a multiple of 256 and IOR exited 0, leading to silent failure and a test that we thought was passing actually failing.

File: src/tests/ftest/harness/pre_launch.py
Comment (ashleypittman) [inline src/tests/ftest/harness/pre_launch.py:21]: This comment should include the name of the script that's supposed to have setup the users correctly, otherwise in years to come someone is going to have to go digging or looking in the git history to see what file was added/edited at the same time.

File: src/tests/ftest/harness/pre_launch.py
Comment (ashleypittman) [inline src/tests/ftest/harness/pre_launch.py:26]: Just for my benefit, when will this test get run?

**Review by ashleypittman — COMMENTED** [review]
Only comment here is should the user/group data be in main.sh or would it be better in user_setup.py itself or a config file.  How does this play with running the tests locally or under ansible.

@knard-intel might like to comment further.

File: src/tests/ftest/user_setup.py
Comment (daltonbohning) [reply src/tests/ftest/user_setup.py:113]: > is there a case where it makes sense to create a group without creating the user

At this time, no. But later it would be easy enough to add an option like `--groups-only`

File: src/tests/ftest/harness/pre_launch.py
Comment (daltonbohning) [reply src/tests/ftest/harness/pre_launch.py:21]: Very good point

File: src/tests/ftest/user_setup.py
Comment (daltonbohning) [reply src/tests/ftest/user_setup.py:125]: Since the name of the script is `user_setup.py`, I changed this to a positional and used `nargs='+',`

File: src/tests/ftest/user_setup.py
Comment (daltonbohning) [reply src/tests/ftest/user_setup.py:110]: For our purposes, we always want a group, but linux doesn't require a group when creating a user. The default is to put the user in a group named after the user

**Comment by daltonbohning** [comment]
> should the user/group data be in main.sh or would it be better in user_setup.py itself or a config file. How does this play with running the tests locally or under ansible.

@ashleypittman I'm not sure. In general, I think our infrastructure is lacking a clear way to handle this

File: src/tests/ftest/harness/pre_launch.py
Comment (ashleypittman) [reply src/tests/ftest/harness/pre_launch.py:26]: Can I ask that you make this into a function so that when we have tests that need this setup they can call the function to verify it's there and will fail correctly if it's not.

File: src/tests/ftest/scripts/main.sh
Comment (daltonbohning) [inline src/tests/ftest/scripts/main.sh:203]: From discussion with Phil, we decided it would be better to move this logic into launch.py so setup is done in other environments as well

File: src/tests/ftest/util/user_utils.py
Comment (ashleypittman) [inline src/tests/ftest/util/user_utils.py:11]: You might not need this pylint annotation now.

File: src/tests/ftest/harness/pre_launch.py
Comment (daltonbohning) [reply src/tests/ftest/harness/pre_launch.py:26]: The users are now created by launch.py, so they will exist when the tests run

Comment (daltonbohning) [reply src/tests/ftest/harness/pre_launch.py:27]: Also added functions in user_utils.py that can be used in test cases to sanity check existence

File: src/tests/ftest/launch.py
Comment (ashleypittman) [inline src/tests/ftest/launch.py:33]: I see what you're trying to do here but I don't think that's the correct option, and I'd rather changes to the import/pylint setup were done separately.

See also this line, the util directory should be added to pylint automatically.
https://github.com/daos-stack/daos/blob/master/utils/cq/daos_pylint.py#L369

Comment (daltonbohning) [reply src/tests/ftest/launch.py:33]: > I don't think that's the correct option

What do you think the correct option is? The only "proper" way I can think of is to convert the entire ftest directory into a python module and install it. The problem is launch.py is one level up from util. And some of the files it imports have other imports from the util directory. And those imports will fail, because the relative import path is one level up.

> I'd rather changes to the import/pylint setup were done separately

Do you mean separate PRs? Wouldn't that introduce a pylint error in the meantime?

> the util directory should be added to pylint automatically

There was somewhere else in the process flagging this, but I don't remember where.

Anyway, these two lines are a partial cherry-pick from https://github.com/daos-stack/daos/pull/10621/files

File: src/tests/ftest/launch.py
Comment (daltonbohning) [inline src/tests/ftest/launch.py:1821]: TODO - remove

**Comment by daltonbohning** [comment]
Work is complete, but this PR conflicts with #10621 - waiting for it to land first

File: src/tests/ftest/launch.py
Comment (ashleypittman) [inline src/tests/ftest/launch.py:2367]: Is this needed?

File: src/tests/ftest/launch.py
Comment (ashleypittman) [inline src/tests/ftest/launch.py:2252]: If this is only called locally then you could rename it to _prepare.

File: src/tests/ftest/harness/launch_setup.py
Comment (ashleypittman) [inline src/tests/ftest/harness/launch_setup.py:2]:
```suggestion
(C) Copyright 2022 Intel Corporation.
```

File: src/tests/ftest/launch.py
Comment (daltonbohning) [reply src/tests/ftest/launch.py:2367]: Yes, because someone **could** override the path to be anywhere else:
https://github.com/daos-stack/daos/blob/8b4fd08e41d1848a2088aba6b43811412447069a/src/tests/ftest/launch.py#L1155-L1156

And `useradd` doesn't create parent directories.

File: src/tests/ftest/launch.py
Comment (daltonbohning) [reply src/tests/ftest/launch.py:2252]: I'll defer to @phender on whether this should be `prepare` or `_prepare`, but generally I agree with you.

File: src/tests/ftest/harness/launch_setup.py
Comment (daltonbohning) [reply src/tests/ftest/harness/launch_setup.py:2]: Thanks. Will fix

File: src/tests/ftest/launch.py
Comment (ashleypittman) [reply src/tests/ftest/launch.py:2367]: Ah, I misread this as a duplicate of the mkdir above but I see it's two different directories now.

File: src/tests/ftest/launch.py
Comment (ashleypittman) [reply src/tests/ftest/launch.py:2252]: If it's only called from self.prepare then it should be _prepare, if it's called from elsewhere then the name is correct.  Given that you change the arguments and only change the same object that tells me it's not called from elsewhere.

Comment (daltonbohning) [reply src/tests/ftest/launch.py:2252]: I think we aren't in the mindset of always naming "private" functions appropriately, but we should be.

Comment (ashleypittman) [reply src/tests/ftest/launch.py:2252]: This is clearly defined in the python docs, it's appropriate to change things to private if you've identified something that can be.

https://docs.python.org/3.8/tutorial/classes.html#private-variables

Comment (daltonbohning) [reply src/tests/ftest/launch.py]: Agree. Done

File: src/tests/ftest/launch.py
Comment (phender) [inline src/tests/ftest/launch.py:1771]: I'd be careful with assuming the expected output is always on the first line, e.g., `stdout[0]`.  A regex here may be safer.  Also note that `result.passed` does not mean that all the hosts returned the same output, that's what `result.homogeneous` indicates.

**Review by phender — CHANGES_REQUESTED** [review]

File: src/tests/ftest/harness/launch_setup.py
Comment (daltonbohning) [reply src/tests/ftest/harness/launch_setup.py]: Done

File: src/tests/ftest/launch.py
Comment (daltonbohning) [reply src/tests/ftest/launch.py]: I added a helper function that robust enough to even handle extraneous output like from SSH, RSA keys, etc.

File: src/tests/ftest/launch.py
Comment (daltonbohning) [reply src/tests/ftest/launch.py]: Fixed

**Review by phender — APPROVED** [review]

File: src/tests/ftest/harness/launch_setup.py
Comment (ashleypittman) [inline src/tests/ftest/harness/launch_setup.py:32]: I can't work this out, get_group_id gives a dict of host/gid but this then takes keys() so it will return a tuple of hosts, which you then call list() on below?

Comment (daltonbohning) [reply src/tests/ftest/harness/launch_setup.py:32]: `get_group_id` returns a dict like `{<gid>: <hosts>}`, so `list(keys())` effectively gives all the unique gids. We don't need the hosts here, but the function returns the dictionary format to be more complete/extensible

Comment (ashleypittman) [reply src/tests/ftest/harness/launch_setup.py:32]: Ah, so it's `gid: host` not `host: gid`.  That makes a lot more sense.

Comment (daltonbohning) [reply src/tests/ftest/harness/launch_setup.py:32]: Right. The main reason being that `hosts` is a `NodeSet`, which isn't hashable

**Review by ashleypittman — APPROVED** [review]

**Comment by daltonbohning** [comment]
Unit tests were skipped since this PR only touches functional tests.
Only VM tests were ran because the changes are verified by the new harness test:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-10521/29/tests

**Comment by daltonbohning** [comment]
@daos-stack/daos-gatekeeper Please use the following commit message on landing:

```
DAOS-11695 test: Add test users to CI env

- Add a --user_create option to launch.py to trigger setting up user accounts from the test yaml
  - launch.py looks for client_users in the yaml
- Add a harness test to verify this functionality, and verify running commands remotely as the created users
- 
```

**Pull request merged**
